### ISSUE TITLE
FOUC on downloads and extensions pages

### ISSUE THREAD

**Issue by dstillman** [zotero/web-components#8]
On the downloads and extensions pages, there are flashes of unstyled content during page loading. E.g., on the extensions page it shows "One button does it all" at the top and a giant gray box that goes off the right side of the screen, creating a horizontal scroll bar. (There are also some weird style changes in the header, so maybe this is a more general issue.) We should add some CSS to hide the components until they're rendered or give them fixed dimensions so that content from below doesn't show up top.

**Comment by flachware** [comment]
This does not look like a classic FOUC to me, but there are several issues:

* The Zotero logo in the header jumps in the base layout, because the `<img>` lacks dimensions. We should add `width="301"` and `height="67"` to the base layout.
* There have been some images in downloads without dimensions, I added them in `web-components`.
* The 'Upgrade Storage' button styles got overridden by the web component syles, I fixed that in `web-components` (also let the layout look jumpy).
* Different `font-smoothing` from the web components was also overriding the base styles, fixed that in `web-components`.
* The `<link>` element in the middle of the markup looks unorthodox to me (and could theoretically cause a FOUC), but seems to be valid and not the cause of the actual issue.
* But the actual issue is, that the react components in extensions and downloads take quite some time until they initialize (about one to several seconds for me), why is that? So we don’t really have unstyled content, all content looks styled to me at any point, but the extensions picker simply comes in quite late, so the 'One button does it all' section is on top first. Also on the downloads page the whole content container is collapsed until the downloads component is initialized. So on the downloads page, a `min-height` of `699px` on `.downloads` (must be in the Zotero styles, not in the web-component styles) would improve the loading behavior. I think an even better solution would be to adapt the base layout to a sticky footer approach, but this would need more work to be done than just setting a min height. On the extensions page we could set a `min-height` of `501px` and `452px` (for `@media (min-width 768px)`) on `#extensions-picker` (must also be in the Zotero styles). Obviously setting a min height is not a perfect solution, it’s a responsive layout, due to the text reflow the elements in question can have various heights.

I’m wondering though if we couldn’t find a generic solution to the problem in order to not end up with setting lots of approximated min-heights all over the place (with an eye to a future website where we have lots of client sided components). Something like a sticky footer layout (so that the layout never collapses) and some logic that waits until all components are initialized and fades them in at once (ev. showing a spinner after some delay).

**Comment by dstillman** [comment]
> So we don’t really have unstyled content, all content looks styled to me at any point, but the extensions picker simply comes in quite late, so the 'One button does it all' section is on top first.

No, I'm actually getting partially unstyled content. It's not a traditional full-page FOUC, but some styles clearly aren't taking effect during the load:

<img width="300" alt="screen shot 2017-06-15 at 3 44 22 pm" src="https://user-images.githubusercontent.com/783082/27198835-e5be315e-51e1-11e7-9562-afdd76833287.png">

<img width="800" alt="screen shot 2017-06-15 at 3 45 44 pm" src="https://user-images.githubusercontent.com/783082/27198874-08e6114c-51e2-11e7-9123-a56c1c35df33.png">

The downloads list is vertical, the "one button" line isn't centered or large, etc.

(I'm taking these in Firefox, which is helpfully extraordinarily slow, but the same things happen in Chrome, just for less time.)

> The <link> element in the middle of the markup looks unorthodox to me (and could theoretically cause a FOUC), but seems to be valid and not the cause of the actual issue.

No, it's totally because of this. If I remove these lines, I get the unstyled content shown above permanently. Why are these here?

> some logic that waits until all components are initialized and fades them in at once (ev. showing a spinner after some delay)

Maybe a separate issue given the unstyled content, but do we want to explore prerendering components server-side with (search-engine-suitable) default values, keeping them hidden (but in layout) initially, and showing them after update?

**Comment by dstillman** [comment]
Fixed by https://github.com/zotero/web-components/commit/36e970eb4ac4a20dccd93d1a7451cd60f164e744

(We still might want to do something about the footer, though.)

**Comment by fcheslack** [comment]
> No, it's totally because of this. If I remove these lines, I get the unstyled content shown above permanently. Why are these here?

Just out of laziness when I set up the demos. Moved to head now, along with the script since it's actually vital to rendering.

> But the actual issue is, that the react components in extensions and downloads take quite some time until they initialize (about one to several seconds for me), why is that?

This might be mostly because it's a dev build when using the demo pages and 'npm start' (though I very rarely get a load that takes several seconds, even in Firefox)

I've added a first (fairly brittle) shot at showing specified content only after the top level react component is mounted.
